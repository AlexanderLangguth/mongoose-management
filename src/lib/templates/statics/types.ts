/**
 * ######################################################################
 * #                                                                    #
 * #                       Do not change the file!                      #
 * #                                                                    #
 * # All changes are deleted the next time the collection data is read. #
 * #                                                                    #
 * ######################################################################
 */

import { IndexOptions } from 'mongodb';
import { SchemaTypeOpts } from 'mongoose';

/**
 * Helpers
 */
type Keys<T> = { [K in keyof T]: T[K] extends never ? never : K }[keyof T];
type Diff<T, U> = T extends U ? never : T;

/**
 * Extracts the type from the function
 */
type virtualFunctionTypeGet<T extends { [k: string]: any }> = T['get'] extends () => void
  ? ReturnType<T['get']>
  : never;
type virtualFunctionTypeSet<T extends { [k: string]: any }> = T['set'] extends (value: infer A) => void ? A : never;

/**
 * List of all types (with the type "never")
 */
type virtualFunctionTypeGetList<T> = { [K in keyof T]: virtualFunctionTypeGet<T[K]> };
type virtualFunctionTypeSetList<T> = { [K in keyof T]: virtualFunctionTypeSet<T[K]> };

/**
 * List of all types (without the type "never")
 */
type virtualTypeGetList<T> = Pick<virtualFunctionTypeGetList<T>, Keys<virtualFunctionTypeGetList<T>>>;
type virtualTypeSetList<T> = Pick<virtualFunctionTypeSetList<T>, Keys<virtualFunctionTypeSetList<T>>>;

/**
 * List of Keys where the object have only "get"
 */
type virtualKeysGetOnly<T> = Diff<keyof virtualTypeGetList<T>, keyof virtualTypeSetList<T>>;

/**
 * List of Keys where without "get only" objects
 */
type virtualKeysBoth<T> = Diff<keyof T, virtualKeysGetOnly<T>>;

export type virtualType<T> = {
  readonly [K in keyof virtualTypeGetList<T> & virtualKeysGetOnly<T>]: virtualTypeGetList<T>[K];
} &
  { [K in keyof virtualTypeGetList<T> & virtualKeysBoth<T>]: virtualTypeGetList<T>[K] } &
  { [K in keyof virtualTypeSetList<T>]: virtualTypeSetList<T>[K] };

/**
 *
 */
export type extendTimestampType = {
  readonly createdAt: Date;
  readonly updatedAt: Date;
};

/**
 *
 */
export type extendIdType = {
  readonly _id: any;
  readonly id: any;
};

/**
 *
 */
export type indexType = { fields: { [k: string]: number }; options?: IndexOptions };

/**
 *
 */
export type schemaExtendType<T extends Record<string, any>> = {
  [K in keyof T]?: T[K] extends Array<schemaExtendType<T>> ? [schemaExtendType<T[K][0]>] : SchemaTypeOpts<any>;
};
