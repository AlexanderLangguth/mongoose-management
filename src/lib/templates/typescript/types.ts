/**
 * ######################################################################
 * #                                                                    #
 * #                       Do not change the file!                      #
 * #                                                                    #
 * # All changes are deleted the next time the collection data is read. #
 * #                                                                    #
 * ######################################################################
 */

namespace Virtual {
  /**
   * Helpers
   */
  type Keys<T> = { [K in keyof T]: T[K] extends never ? never : K }[keyof T];
  type Diff<T, U> = T extends U ? never : T;

  /**
   * Extracts the type from the function
   */
  type functionTypeGet<T extends { [k: string]: any }> = T['get'] extends Function ? ReturnType<T['get']> : never;
  type functionTypeSet<T extends { [k: string]: any }> = T['set'] extends (value: infer A) => void ? A : never;

  /**
   * List of all types (with the type "never")
   */
  type functionTypeGetList<T> = { [K in keyof T]: functionTypeGet<T[K]> };
  type functionTypeSetList<T> = { [K in keyof T]: functionTypeSet<T[K]> };

  /**
   * List of all types (without the type "never")
   */
  type typeGetList<T> = Pick<functionTypeGetList<T>, Keys<functionTypeGetList<T>>>;
  type typeSetList<T> = Pick<functionTypeSetList<T>, Keys<functionTypeSetList<T>>>;

  /**
   * List of Keys where the object have only "get"
   */
  type keysGetOnly<T> = Diff<keyof typeGetList<T>, keyof typeSetList<T>>;

  /**
   * List of Keys where without "get only" objects
   */
  type keysBoth<T> = Diff<keyof T, keysGetOnly<T>>;

  export type types<T> = { readonly [K in keyof typeGetList<T> & keysGetOnly<T>]: typeGetList<T>[K] } &
    { [K in keyof typeGetList<T> & keysBoth<T>]: typeGetList<T>[K] } &
    { [K in keyof typeSetList<T>]: typeSetList<T>[K] };
}

export type virtualType<T> = Virtual.types<T>;

export type extendTimestampType = {
  readonly createdAt: Date;
  readonly updatedAt: Date;
};

export type extendIdType = {
  readonly _id: any;
  readonly id: any;
};
