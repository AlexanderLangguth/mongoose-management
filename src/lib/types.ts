import { Separator } from 'inquirer';

import Prompts from './prompts';
import Storage from './storage';

import Collection from './cli/sequences/collection';
import Collections from './cli/sequences/collections';
import Column from './cli/sequences/column';
import Index from './cli/sequences/index';

import CollectionAction from './cli/actions/collection';
import ColumnAction from './cli/actions/column';
import GroupAction from './cli/actions/group';
import IndexAction from './cli/actions/index';

import Create from './template/create';

import * as mongoTypes from './mongo';

/**
 *
 */
export type objType<K extends string | number | symbol = string, T = any> = Record<K, T>;

/**
 *
 */
export type cliOptionsType = {
  prompts: Prompts;
  storage: Storage;

  // sequences
  Collections: typeof Collections;
  Collection: typeof Collection;
  Column: typeof Column;
  Index: typeof Index;

  // actions
  actionGroup: GroupAction;
  actionCollection: CollectionAction;
  actionColumn: ColumnAction;
  actionIndex: IndexAction;

  createTemplate: Create;

  data: dataType;
};

/**
 *
 */
export type schemaType = keyof typeof mongoTypes.schemaTypes;

/**
 *
 */
export type schemaNormalType = keyof typeof mongoTypes.schemaTypesNormal;

/**
 *
 */
export type schemaSpecialType = keyof typeof mongoTypes.schemaTypesSpecial;

/**
 *
 */
export type schemaIndexType = keyof typeof mongoTypes.schemaIndexTypes;

/**
 *
 */
export type dataType = {
  groups: dataGroupType[];
};

/**
 *
 */
export type dataGroupType = {
  path: string;
  collections: dataCollectionType[];
};

/**
 *
 */
export type dataCollectionType = {
  name: string;
  columns: dataColumnType[];
  indexes: dataIndexType[];
};

/**
 *
 */
export type dataColumnType = {
  name: string;
  type: schemaType;

  required: boolean;
  default?: string;

  // string
  lowercase?: boolean;
  uppercase?: boolean;
  trim?: boolean;
  match?: string;
  enum?: string;
  minlength?: number;
  maxlength?: number;

  // number
  min?: number;
  max?: number;

  // internal values
  subType?: dataColumnArrayType;
  subColumns?: dataColumnType[];
  showType?: string;
  readonly?: boolean;
};

/**
 *
 */
export type dataColumnArrayType = {
  type: schemaNormalType;
  subType?: dataColumnArrayType;
};

/**
 *
 */
export type dataIndexType = {
  name: string;
  columns: { [k: string]: dataIndexColumnValueType };
  properties: {
    unique?: boolean;
    sparse?: boolean;
    /*
    expires?: number;
    artialFilterExpression?: any;
    collation?: {
      locale: string;
      strength: number;
    };
    */
  };
  // internal values
  readonly?: boolean;
  mode?: schemaIndexType;
  type?: dataIndexColumnValueType;
};

/**
 *
 */
export type dataColumnGroupType = {
  parent?: dataColumnGroupType;
  column: dataColumnType;
};

/**
 *
 */
export type dataIndexColumnValueType = 1 | -1 | 'text' | 'hashed' | '2dsphere';

/**
 *
 */
export type choicesType<R> = choiceType<R> | InstanceType<typeof Separator>;

/**
 *
 */
export type choiceType<R> = choiceListType<choiceValueType<R>>;

/**
 *
 */
export type choiceListType<T> = {
  name: string;
  value: T;
  short: string;
};

/**
 *
 */
export type choiceValueType<R> = {
  type?: 'create' | 'back' | 'exit' | 'save' | 'edit' | 'remove' | 'write';
  value?: R;
};

/**
 *
 */
export type choiceValueSubType =
  | {
      type: 'column';
      data?: dataColumnType;
    }
  | {
      type: 'index';
      data?: dataIndexType;
    };

/**
 *
 */
export type templateTypesType = 'index' | 'documents' | 'interfaces' | 'models' | 'repositories';

/**
 *
 */
export type templateCollectionNamesType = {
  collectionNameLower: string;
  collectionNameUpper: string;
  collectionNameRaw: string;
};

/**
 *
 */
export type templateCollectionType = {
  interfaceName: string;

  schemaDefinitions: string;
  SchemaIndexes: string;
  schemaTypes: string;
} & templateCollectionNamesType;
