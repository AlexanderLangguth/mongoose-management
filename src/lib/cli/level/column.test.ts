jest.mock('../../prompts');
jest.mock('../../storage');
jest.mock('../../template/create');
jest.mock('../dataset/collection');
jest.mock('../dataset/column');
jest.mock('../dataset/index');
jest.mock('../menu/column');
jest.mock('../prompts/column');

import Prompts from '../../prompts';
import Storage from '../../storage';
import Creater from '../../template/create';

import CollectionDataset from '../dataset/collection';
import ColumnDataset from '../dataset/column';
import IndexDataset from '../dataset/index';

import ColumnMenu from '../menu/column';

import promptsColumn from '../prompts/column';

import ColumnLevel from './column';

import { levelOptionsType } from '../../types';

/**
 *
 */
describe('Check the ColumnLevel class', () => {
  let level: any;
  let dataset: any;
  let prompts: any;
  let options: levelOptionsType;

  /**
   *
   */
  beforeEach(() => {
    const storage = new (Storage as any)();
    const creater = new (Creater as any)();
    prompts = new (Prompts as any)();
    dataset = new (ColumnDataset as any)();
    options = { prompts, storage, creater };

    level = new ColumnLevel(dataset, options);
  });

  /**
   *
   */
  test('initialize the class', () => {
    expect(level).toBeInstanceOf(ColumnLevel);

    expect(level.dataset).toBe(dataset);
    expect(level.menu).toBeInstanceOf(ColumnMenu);
    expect(level.prompts).toBe(prompts);
    expect(level.options).toBe(options);

    expect(level.promptCreate).toEqual(expect.any(Function));
    expect(level.promptEdit).toEqual(expect.any(Function));

    expect(level.promptCreate).toBe(promptsColumn);
    expect(level.promptEdit).toBe(promptsColumn);
  });

  /**
   *
   */
  test.each([['array'], ['object']])(
    'it should be promptsColumn() called when create() is called with type "%s"',
    async (action) => {
      const mock = new (ColumnDataset as any)();

      mock.get = jest.fn().mockReturnValue(action);

      (promptsColumn as jest.Mock).mockResolvedValue(mock);

      const result = await level.create('create');

      expect(result).toBe(mock);

      expect(promptsColumn).toHaveBeenCalledTimes(1);
      expect(promptsColumn).toHaveBeenCalledWith(prompts, dataset);
    },
  );

  /**
   *
   */
  test('it should be promptsColumn() called when create() is called with type "string"', async () => {
    const mock = new (ColumnDataset as any)();

    mock.get = jest.fn().mockReturnValue('string');

    (promptsColumn as jest.Mock).mockResolvedValue(mock);

    const result = await level.create('create');

    expect(result).toBe(undefined);

    expect(promptsColumn).toHaveBeenCalledTimes(1);
    expect(promptsColumn).toHaveBeenCalledWith(prompts, dataset);
  });

  /**
   *
   */
  test('it should be remove dataset when remove() is called with confirmation', async () => {
    const mockColumn = new (ColumnDataset as any)();
    const mockIndex = new (IndexDataset as any)();
    const mockIndexOther1 = new (IndexDataset as any)();
    const mockIndexOther2 = new (IndexDataset as any)();
    const mockCollection = new (CollectionDataset as any)();

    mockColumn.getColumns.mockReturnValue([]);
    mockColumn.getName.mockReturnValue('subcolumn');
    mockColumn.getFullname.mockReturnValue('columns.subcolumn');
    mockColumn.getIndex.mockReturnValue(mockIndex);
    mockColumn.getCollection.mockReturnValue(mockCollection);

    mockCollection.getIndexes.mockReturnValue([mockIndexOther1, mockIndex, mockIndexOther2]);

    mockIndex.getColumns.mockReturnValue({ 'columns.subcolumn': 'text' });
    mockIndex.getName.mockReturnValue('main-index');
    mockIndexOther1.getColumns.mockReturnValue({ b: 1, c: -1 });
    mockIndexOther1.getName.mockReturnValue('other-index1');
    mockIndexOther2.getColumns.mockReturnValue({ a: 1, b: 1 });
    mockIndexOther2.getName.mockReturnValue('other-index2');

    prompts.remove.mockResolvedValue(true);

    const result = await level.remove(mockColumn);

    expect(result).toBe(false);

    expect(mockColumn.getColumns).toHaveBeenCalledTimes(1);
    expect(mockColumn.getName).toHaveBeenCalledTimes(1);
    expect(mockColumn.getFullname).toHaveBeenCalledTimes(1);
    expect(mockColumn.getFullname).toHaveBeenCalledWith(false, false);
    expect(mockColumn.getIndex).toHaveBeenCalledTimes(1);
    expect(mockColumn.getCollection).toHaveBeenCalledTimes(1);
    expect(mockColumn.remove).toHaveBeenCalledTimes(1);

    expect(mockCollection.getIndexes).toHaveBeenCalledTimes(1);

    expect(mockIndex.getColumns).toHaveBeenCalledTimes(0);
    expect(mockIndex.getName).toHaveBeenCalledTimes(0);

    expect(mockIndexOther1.getColumns).toHaveBeenCalledTimes(1);
    expect(mockIndexOther1.getName).toHaveBeenCalledTimes(0);

    expect(mockIndexOther2.getColumns).toHaveBeenCalledTimes(1);
    expect(mockIndexOther2.getName).toHaveBeenCalledTimes(0);

    expect(prompts.remove).toHaveBeenCalledTimes(1);
  });

  /**
   *
   */
  test('it should be not remove dataset when remove() is called without confirmation', async () => {
    const mockColumn = new (ColumnDataset as any)();
    const mockIndex = new (IndexDataset as any)();
    const mockIndexOther1 = new (IndexDataset as any)();
    const mockIndexOther2 = new (IndexDataset as any)();
    const mockCollection = new (CollectionDataset as any)();

    mockColumn.getColumns.mockReturnValue([]);
    mockColumn.getName.mockReturnValue('subcolumn');
    mockColumn.getFullname.mockReturnValue('columns.subcolumn');
    mockColumn.getIndex.mockReturnValue(mockIndex);
    mockColumn.getCollection.mockReturnValue(mockCollection);

    mockCollection.getIndexes.mockReturnValue([mockIndexOther1, mockIndex, mockIndexOther2]);

    mockIndex.getColumns.mockReturnValue({ 'columns.subcolumn': 'text' });
    mockIndex.getName.mockReturnValue('main-index');
    mockIndexOther1.getColumns.mockReturnValue({ b: 1, c: -1 });
    mockIndexOther1.getName.mockReturnValue('other-index1');
    mockIndexOther2.getColumns.mockReturnValue({ a: 1, b: 1 });
    mockIndexOther2.getName.mockReturnValue('other-index2');

    prompts.remove.mockResolvedValue(false);

    const result = await level.remove(mockColumn);

    expect(result).toBe(true);

    expect(mockColumn.getColumns).toHaveBeenCalledTimes(1);
    expect(mockColumn.getName).toHaveBeenCalledTimes(1);
    expect(mockColumn.getFullname).toHaveBeenCalledTimes(1);
    expect(mockColumn.getFullname).toHaveBeenCalledWith(false, false);
    expect(mockColumn.getIndex).toHaveBeenCalledTimes(1);
    expect(mockColumn.getCollection).toHaveBeenCalledTimes(1);
    expect(mockColumn.remove).toHaveBeenCalledTimes(0);

    expect(mockCollection.getIndexes).toHaveBeenCalledTimes(1);

    expect(mockIndex.getColumns).toHaveBeenCalledTimes(0);
    expect(mockIndex.getName).toHaveBeenCalledTimes(0);

    expect(mockIndexOther1.getColumns).toHaveBeenCalledTimes(1);
    expect(mockIndexOther1.getName).toHaveBeenCalledTimes(0);

    expect(mockIndexOther2.getColumns).toHaveBeenCalledTimes(1);
    expect(mockIndexOther2.getName).toHaveBeenCalledTimes(0);

    expect(prompts.remove).toHaveBeenCalledTimes(1);
  });

  /**
   *
   */
  test('it should be throw an error when remove() is called with subcolumns', async () => {
    const mock = new (ColumnDataset as any)();
    const mockSubColumn1 = new (ColumnDataset as any)();
    const mockSubColumn2 = new (ColumnDataset as any)();

    mock.getColumns.mockReturnValue([mockSubColumn1, mockSubColumn2]);

    expect.assertions(2);
    try {
      await level.remove(mock);
    } catch (err) {
      expect(err).toBeInstanceOf(Error);
      expect(err.message).toBe('There are still subcolumns. These must be deleted first!');
    }
  });

  /**
   *
   */
  test('it should be throw an error when remove() is called with indexes found', async () => {
    const mockColumn = new (ColumnDataset as any)();
    const mockIndex = new (IndexDataset as any)();
    const mockIndexOther1 = new (IndexDataset as any)();
    const mockIndexOther2 = new (IndexDataset as any)();
    const mockCollection = new (CollectionDataset as any)();

    mockColumn.getColumns.mockReturnValue([]);
    mockColumn.getName.mockReturnValue('subcolumn');
    mockColumn.getFullname.mockReturnValue('columns.subcolumn');
    mockColumn.getIndex.mockReturnValue(mockIndex);
    mockColumn.getCollection.mockReturnValue(mockCollection);

    mockCollection.getIndexes.mockReturnValue([mockIndexOther1, mockIndex, mockIndexOther2]);

    mockIndex.getColumns.mockReturnValue({ 'columns.subcolumn': 'text' });
    mockIndex.getName.mockReturnValue('main-index');
    mockIndexOther1.getColumns.mockReturnValue({ b: 1, c: -1 });
    mockIndexOther1.getName.mockReturnValue('other-index1');
    mockIndexOther2.getColumns.mockReturnValue({ 'columns.subcolumn': 1, b: 1 });
    mockIndexOther2.getName.mockReturnValue('other-index2');

    expect.assertions(17);
    try {
      await level.remove(mockColumn);
    } catch (err) {
      expect(err).toBeInstanceOf(Error);
      expect(err.message).toBe(
        [
          'Indexes still exist for the column. These must be deleted first!',
          '',
          'The following indexes contain the column:',
          '- other-index2',
        ].join('\n'),
      );

      expect(mockColumn.getColumns).toHaveBeenCalledTimes(1);
      expect(mockColumn.getName).toHaveBeenCalledTimes(0);
      expect(mockColumn.getFullname).toHaveBeenCalledTimes(1);
      expect(mockColumn.getFullname).toHaveBeenCalledWith(false, false);
      expect(mockColumn.getIndex).toHaveBeenCalledTimes(1);
      expect(mockColumn.getCollection).toHaveBeenCalledTimes(1);
      expect(mockColumn.remove).toHaveBeenCalledTimes(0);

      expect(mockCollection.getIndexes).toHaveBeenCalledTimes(1);

      expect(mockIndex.getColumns).toHaveBeenCalledTimes(0);
      expect(mockIndex.getName).toHaveBeenCalledTimes(0);

      expect(mockIndexOther1.getColumns).toHaveBeenCalledTimes(1);
      expect(mockIndexOther1.getName).toHaveBeenCalledTimes(0);

      expect(mockIndexOther2.getColumns).toHaveBeenCalledTimes(1);
      expect(mockIndexOther2.getName).toHaveBeenCalledTimes(1);

      expect(prompts.remove).toHaveBeenCalledTimes(0);
    }
  });

  /**
   *
   */
  test('it should be create ConsumerLevel when show() is called', async () => {
    const mock = new (ColumnDataset as any)();

    ColumnLevel.prototype.exec = jest.fn();

    await level.show(mock);

    expect(ColumnLevel.prototype.exec).toHaveBeenCalledTimes(1);
    expect(ColumnLevel.prototype.exec).toHaveBeenCalledWith();
  });
});
