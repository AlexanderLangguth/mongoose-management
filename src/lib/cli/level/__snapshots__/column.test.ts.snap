// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check the ColumnLevel class it should be throw an error when remove() is called with a reference 1`] = `
"This nested schema is referenced by other column(s):

- [1mc1[22m.cName

The references must first be deleted"
`;

exports[`Check the ColumnLevel class it should be throw an error when remove() is called with indexes found 1`] = `
"Indexes still exist for the column. These must be deleted first!

The following indexes contain the column:
- name_unique"
`;
